docs/DesignDocs/CustomAttestation.md

```diff
@@ -0,0 +1,849 @@
+ 0x4f, 0xd4,                            \
+ 0x86, 0x63,                            \
+ 0x42, 0x1e, 0x3a, 0x57, 0xa0, 0xa4}    \
+}
+```
+
+`my_plugin.c`
+```C
+/* Plugin implementation functions here. */
+static oe_result_t my_plugin_on_register(...) {...}
+
+/* Setting up the plugin struct. */
+oe_attestation_plugin_t my_plugin = {
+ /* Plugin UUID. */
+ .format_id = MY_PLUGIN_UUID,
+
+  /* Plugin functions. */
+ .on_register = my_plugin_on_register,
+  /* Rest of the plugin functions. */
+};
+
+/* Implement helper initialization function. */
+oe_attestation_plugin_t* my_plugin() {
+    return &my_plugin;
+}
+```
+
+They can then compile their code in the standard way for building Open Enclave
+enclave and host applications.
+
+Plugin consumers will use the new "plugin aware" APIs like
+`oe_get_attestation_evidence`. Plugin consumers will use the APIs like this:
+
+use_plugin.c
+```C
+#include <my_plugin.h> // For my_plugin() helper function and UUID.
+
+/* Register plugin. Send the config data if necessary. */
+oe_register_plugin(my_plugin(), my_config_data, my_config_data_size);
+
+/* Get evidence. */
+oe_get_attestation_evidence(
+    MY_PLUGIN_UUID,
+    my_custom_claims,
+    my_custom_claims_size,
+    &evidence,
+    &evidence_size);
+
+/* Get collaterals. */
+oe_get_attestation_collateral(
+    MY_PLUGIN_UUID,
+    &collaterals,
+    &collaterals_size);
+
+/* Verify evidence. Can check the claims if desired. */
+oe_verify_attestaton_evidence(
+    evidence,
+    evidence_size,
+    collaterals,
+    collaterals_size,
+    &claims,
+    &claims_size);
+
+/* Unregister plugin. */
+oe_unregister_plugin(my_plugin());
+
+```
+
+The plugin user can now link in the plugin to build their app:
+
+```bash
+gcc -o my_app use_plugin.o my_plugin.o ...
+```
+
+### Remote attestation collaterals
+
+There are 2 user scenarios.  Examples shows running from an enclave, running from the host should be similar.
+
+#### 1. User does not specify the collaterals:
+```c
+...
+
+// Get evidence
+result = oe_get_attestation_evidence(
+    MY_PLUGIN_UUID,
+    NULL, // custom_claims
+    0,
+    NULL, // opt_params
+    0,
+    &evidence,
+    &evidence_size);
+
+// Verify report without collaterals
+result = oe_verify_attestation_evidence(
+            evidence,
+            evidence_size,
+            NULL, // collaterals
+            0,    // collaterals_size
+            NULL, // verification_params
+            0,    // verification_params_size
+            &claims,
+            &claims_size);
+...
+
+```
+
+#### 2. User specifies collaterals:
+```c
+...
+
+// Get evidence
+result = oe_get_attestation_evidence(
+    MY_PLUGIN_UUID,
+    NULL, // custom_claims
+    0,
+    NULL, // opt_params
+    0,
+    &evidence,
+    &evidence_size);
+
+// Get collaterals
+result = oe_get_attestation_collaterals(
+    MY_PLUGIN_UUID,
+    &collaterals,
+    &collaterals_size);
+
+// Verify evidence with collateral
+result = oe_verify_attestation_evidence(
```

Comment:
This example implies that the same entity calls oe_get_attestation_evidence as calls oe_verify_attestation_evidence, which I dont think is intended as the recommended behavior.  Instead shouldn't there be two example snippets that work together?

Reply:
Fixed.